WooCommerce/Classes/ViewRelated/Orders/Order Creation/Synchronizer/RemoteOrderSynchronizer.swift

```diff
@@ -105,58 +105,156 @@ private extension RemoteOrderSynchronizer {
             .merge(with: setShipping.map { _ in () })
             .merge(with: setFee.map { _ in () })
+            .debounce(for: 1, scheduler: DispatchQueue.main) // Group & wait for 1s since the last signal was emitted.
+            .compactMap { [weak self] in
+                guard let self = self else { return nil }
+                return SyncOperation(order: self.order) // Imperative `withLatestFrom` as it appears to have bugs when assigning a new order value.
+            }
+            .share()
             .eraseToAnyPublisher()
 
+        bindOrderCreation(trigger: syncTrigger, forceUpdateSignal: forceUpdateSignal)
+        bindOrderUpdate(trigger: syncTrigger, forceUpdateSignal: forceUpdateSignal)
+    }
+
+    /// Binds the provided `trigger` and creates an order when needed(order does not exists remotely).
+    /// Keeps track of subsequent update triggers to send a signal to the provided `forceUpdateSignal`.
+    ///
+    func bindOrderCreation(trigger: AnyPublisher<SyncOperation, Never>, forceUpdateSignal: PassthroughSubject<SyncOperation, Never>) {
+        // Stores the latest request. Needed to validate if an update needs to be fired after the order is created.
+        var latestRequest: SyncOperation?
+
         // Creates a "draft" order if the order has not been created yet.
-        syncTrigger
-            .debounce(for: 0.5, scheduler: DispatchQueue.main) // Group & wait for 0.5s since the last signal was emitted.
-            .withLatestFrom(orderPublisher)
-            .filter { _, order in
-                order.orderID == .zero // Only continue if the order has not been created.
+        trigger
+            .handleEvents(receiveOutput: { request in
+                latestRequest = request // Assign latest request to further evaluation.
+            })
+            .filter { // Only continue if the order has not been created.
+                $0.order.orderID == .zero
+            }
+            .flatMap(maxPublishers: .max(1)) { [weak self] request -> AnyPublisher<SyncOperation, Error> in // Only allow one request at a time.
+                guard let self = self else { return Empty().eraseToAnyPublisher() }
+                self.state = .syncing
+                return self.createOrderRemotely(request)
             }
-            .flatMap(maxPublishers: .max(1)) { [weak self] (_, order) -> AnyPublisher<Order, Error> in // Only allow one request at a time.
+            .catch { [weak self] error -> AnyPublisher<SyncOperation, Never> in // When an error occurs, update state & finish.
+                self?.state = .error(error)
+                return Empty().eraseToAnyPublisher()
+            }
+            .sink { [weak self] response in
+                // If there are no pending update requests, update state & order.
+                if response.id == latestRequest?.id {
+                    self?.state = .synced
+                    self?.order = response.order
```

Comment:
I encountered an issue with the way this check is being done: Although this checks for any saved changes to the order, it doesn't detect if new order data has been entered in the UI but not saved. For example:

1. Enter a shipping line and tap Done.
2. Notice a request is made to create the new order remotely.
3. Tap on "Add Fees" and enter a fee amount. While on that screen, the first response is received and the entered data disappears.

https://user-images.githubusercontent.com/8658164/155539212-434e5b8b-c294-4618-9d56-7c691e7454c7.mp4

The UI also becomes unresponsive after that. I'm not entirely sure but I'm guessing it's related to how we handle the view models when the order changes.

One idea to handle this: we could notify the synchronizer when the user navigates away from the `NewOrder` screen (i.e. to add some new data to the order) and not take any action on pending responses until the user has returned to the `NewOrder` screen, in case they are triggering a new update request. I'm not sure if there's a better way to handle that scenario, though.

Reply:
Good catch, My assumption is that this is probably due to how we are handling bindings on the fees screen. I think the fees screen should not read/bind from the order on the main screen.

As this needs further investigation, here is the follow-up issue.

- https://github.com/woocommerce/woocommerce-ios/issues/6306